File: 186. Reverse Words in a String II.py
class Solution:
    '''
    First reverse all char, then reverse word
    '''
    def reverseWords(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        if not s:
            return
        sp = 0
        self.rev(0, len(s)-1, s)
        for i in range(len(s)):
            if s[i] == ' ':
                self.rev(sp, i-1, s)
                sp = i+1
        self.rev(sp, len(s)-1, s)
        
    def rev(self, st, en, s):
        for i in range((en-st+1)//2):
            s[st+i], s[en-i] = s[en-i], s[st+i]
        return

File: 186. Reverse Words in a String II.java
class Solution {
    public void reverseWords(char[] s) {
        reverse(s, 0, s.length-1);
        int w = 0;
        for (int i=0; i<s.length; i++) {
            if (s[i] == ' ') {
                reverse(s, w, i-1);
                w = i+1;
            }
        }
        reverse(s, w, s.length-1);
    }
    private void reverse(char[] s, int start, int end) {
        while (start < end) {
            char temp = s[start];
            s[start] = s[end];
            s[end] = temp;
            start++; end--;
        }
    }
}